pkg/pillar/cmd/zedagent/handlentp.go

```diff
@@ -0,0 +1,303 @@
+// Copyright (c) 2024 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package zedagent
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"os"
+	"path"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+
+	chrony "github.com/facebook/time/ntp/chrony"
+	"github.com/lf-edge/eve-api/go/info"
+	"github.com/lf-edge/eve/pkg/pillar/flextimer"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	"google.golang.org/protobuf/proto"
+	timestamppb "google.golang.org/protobuf/types/known/timestamppb"
+)
+
+const (
+	// Path to unix chrony socket
+	unixChronydPath = "/run/chrony/chronyd.sock"
+)
+
+// Run a periodic post of the NTP sources information.
+func ntpSourcesTimerTask(ctx *zedagentContext, handleChannel chan interface{},
+	triggerNTPSourcesInfo chan destinationBitset) {
+
+	// Ticker for periodic publishing NTP sources to the controller.
+	globalInterval := ctx.globalConfig.GlobalValueInt(types.NTPSourcesInterval)
+	interval := time.Duration(globalInterval) * time.Second
+	max := float64(interval)
+	min := max * 0.3
+	ticker := flextimer.NewRangeTicker(time.Duration(min), time.Duration(max))
+
+	// Return handles to the caller.
+	handleChannel <- ticker
+
+	wdName := agentName + "-ntp"
+
+	// Run a periodic timer so we always update StillRunning
+	stillRunning := time.NewTicker(25 * time.Second)
+	ctx.ps.StillRunning(wdName, warningTime, errorTime)
+	ctx.ps.RegisterFileWatchdog(wdName)
+
+	for {
+		select {
+		case <-ticker.C:
+			publishNTPSources(ctx, wdName, ControllerDest)
+		case dest := <-triggerNTPSourcesInfo:
+			publishNTPSources(ctx, wdName, dest)
+		case <-stillRunning.C:
+		}
+		ctx.ps.StillRunning(wdName, warningTime, errorTime)
+	}
+}
+
+// Called when globalConfig changes.
+// Assumes that the caller has verifier that the interval has changed.
+func updateNTPSourcesTimer(ctx *getconfigContext, globalInterval uint32) {
+	if ctx.ntpSourcesTickerHandle == nil {
+		log.Warnf("updateNTPSourcesTimer: ticker is still nil")
+		return
+	}
+	interval := time.Duration(globalInterval) * time.Second
+	log.Functionf("updateNTPSourcesTimer: interval change to %v", interval)
+	max := float64(interval)
+	min := max * 0.3
+	flextimer.UpdateRangeTicker(ctx.ntpSourcesTickerHandle,
+		time.Duration(min), time.Duration(max))
+	// Force an immediate timeout since timer could have decreased.
+	flextimer.TickNow(ctx.ntpSourcesTickerHandle)
+}
+
+func publishNTPSources(ctx *zedagentContext, wdName string,
+	dest destinationBitset) {
+	info := getNTPSourcesInfo(ctx)
+	if info == nil {
+		// Not available.
+		return
+	}
+	start := time.Now()
+	publishNTPSourcesToDest(ctx, info, dest)
+	ctx.ps.CheckMaxTimeTopic(wdName, "publishNTPSources", start,
+		warningTime, errorTime)
+}
+
+func publishNTPSourcesToDest(ctx *zedagentContext,
+	infoNTPSources *info.ZInfoNTPSources, dest destinationBitset) {
+
+	// TODO: we don't support LPS
+	if (dest &^ LPSDest) == 0 {
```

Comment:
This is what (at least makes it look like) we don't support LOC. Is there a different dest for LPS and LOC?